File: to-fhir-cs/ToFhir/TransactionBuilder.cs
using System.Security.Cryptography;
using System.Text;
using Hl7.Fhir.Model;

namespace ToFhir;

/// <summary>
/// Builder for creating FHIR transaction bundles. By default, using the update-as-create approach.
/// </summary>
public sealed class TransactionBuilder
{
    private const string DataOperationSystem =
        "http://terminology.hl7.org/CodeSystem/v3-DataOperation";
    private const string ParticipantTypeSystem =
        "http://terminology.hl7.org/CodeSystem/provenance-participant-type";
    private const string ParticipationTypeSystem =
        "http://terminology.hl7.org/CodeSystem/v3-ParticipationType";

    private readonly List<Resource> _resources = [];
    private readonly List<ResourceReference> _resourcesToDelete = [];

    private Bundle.BundleType _bundleType = Bundle.BundleType.Transaction;
    private string? _bundleId;
    private string? _fullUrlBase;
    private bool _failOnDuplicateEntries;
    private ResourceReference? _provenanceWho;
    private ResourceReference? _provenanceWhat;
    private Device? _provenanceDevice;

    private bool IsProvenanceEnabled => _provenanceWho is not null && _provenanceWhat is not null;

    /// <summary>Sets the bundle type.</summary>
    /// <param name="bundleType">The type of bundle to build.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder WithType(Bundle.BundleType bundleType)
    {
        _bundleType = bundleType;
        return this;
    }

    /// <summary>
    /// Configures the builder to throw if multiple resources with the same id are added when the
    /// transaction is built.
    /// </summary>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder FailOnDuplicateEntries()
    {
        _failOnDuplicateEntries = true;
        return this;
    }

    /// <summary>Adds a FHIR resource to the transaction bundle.</summary>
    /// <param name="resource">The FHIR resource to add to the bundle.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder AddEntry(Resource resource)
    {
        ArgumentNullException.ThrowIfNull(resource);
        _resources.Add(resource);
        return this;
    }

    /// <summary>Adds FHIR resources to the transaction bundle.</summary>
    /// <param name="resources">The FHIR resources to add to the bundle.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder AddEntries(params Resource[] resources) =>
        AddEntries((IEnumerable<Resource>)resources);

    /// <summary>Adds a sequence of FHIR resources to the transaction bundle.</summary>
    /// <param name="resources">The FHIR resources to add to the bundle.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder AddEntries(IEnumerable<Resource> resources)
    {
        ArgumentNullException.ThrowIfNull(resources);
        _resources.AddRange(resources);
        return this;
    }

    /// <summary>Adds a reference to a resource that should be deleted as part of the transaction.</summary>
    /// <param name="resource">A reference to the resource to delete.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder AddDeleteEntry(ResourceReference resource)
    {
        ArgumentNullException.ThrowIfNull(resource);
        _resourcesToDelete.Add(resource);
        return this;
    }

    /// <summary>Adds references to resources that should be deleted as part of the transaction.</summary>
    /// <param name="resources">References to the resources to delete.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder AddDeleteEntries(params ResourceReference[] resources) =>
        AddDeleteEntries((IEnumerable<ResourceReference>)resources);

    /// <summary>Adds a sequence of references to resources that should be deleted.</summary>
    /// <param name="resources">References to the resources to delete.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder AddDeleteEntries(IEnumerable<ResourceReference> resources)
    {
        ArgumentNullException.ThrowIfNull(resources);
        _resourcesToDelete.AddRange(resources);
        return this;
    }

    /// <summary>Sets the id of the bundle.</summary>
    /// <param name="id">The id to set for the bundle.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder WithId(string id)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        _bundleId = id;
        return this;
    }

    /// <summary>
    /// Sets an absolute base URL used to build each entry's <c>fullUrl</c> as
    /// <c>&lt;baseUrl&gt;/&lt;ResourceType&gt;/&lt;id&gt;</c>, making the fullUrl absolute as FHIR
    /// requires. Plain <c>ResourceType/id</c> references used elsewhere in the bundle (e.g. in a
    /// <c>Provenance.target</c>) still resolve correctly against it, since those match by the tail
    /// of a hierarchical fullUrl. The base URL does not need to be a real, dereferenceable server
    /// endpoint.
    /// </summary>
    /// <param name="baseUrl">
    /// An absolute base URL, e.g. <c>https://example.org/fhir</c> (a trailing slash is optional).
    /// </param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder WithFullUrlBase(string baseUrl)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(baseUrl);
        _fullUrlBase = baseUrl.TrimEnd('/');
        return this;
    }

    /// <summary>
    /// Enables the inclusion of a Provenance resource in the transaction bundle, with the specified
    /// <c>Provenance.agent.who</c> and <c>Provenance.entity.what</c> references. If the bundle
    /// contains both delete and update/create entries, two Provenance resources will be included.
    ///
    /// <para>The <c>Provenance.id</c> is built from the hash of the <paramref name="who"/> and
    /// <paramref name="what"/> references.</para>
    /// </summary>
    /// <param name="who">
    /// A reference to the agent responsible for the transformation or deletion. This is typically
    /// the transformation service itself.
    /// </param>
    /// <param name="what">A reference to the resource that is the source of the transformation.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder WithProvenance(ResourceReference who, ResourceReference what)
    {
        ArgumentNullException.ThrowIfNull(who);
        ArgumentNullException.ThrowIfNull(what);
        _provenanceWho = who;
        _provenanceWhat = what;
        return this;
    }

    /// <summary>
    /// Enables the inclusion of a Provenance resource in the transaction bundle, using the given
    /// Device as the <c>Provenance.agent.who</c>.
    /// </summary>
    /// <param name="device">
    /// The Device resource representing the agent responsible for the transformation or deletion.
    /// The resource is automatically added to the bundle and referenced in the
    /// <c>Provenance.agent.who</c> element.
    /// </param>
    /// <param name="what">A reference to the resource that is the source of the transformation.</param>
    /// <returns>This builder instance for chaining.</returns>
    public TransactionBuilder WithProvenance(Device device, ResourceReference what)
    {
        ArgumentNullException.ThrowIfNull(device);
        ArgumentNullException.ThrowIfNull(what);
        _provenanceDevice = device;
        _provenanceWho = ReferenceUtils.CreateReferenceTo(device);
        _provenanceWhat = what;
        return this;
    }

    /// <summary>Builds and returns a FHIR Bundle with the configured type.</summary>
    /// <returns>A new Bundle instance with the configured type.</returns>
    /// <exception cref="InvalidOperationException">
    /// <see cref="FailOnDuplicateEntries"/> is enabled and duplicate resource ids were found.
    /// </exception>
    public Bundle Build()
    {
        var bundle = CreateBundle(_bundleType, _bundleId);

        AddDataEntries(bundle);

        if (IsProvenanceEnabled)
        {
            if (_provenanceDevice is not null && !_resources.Contains(_provenanceDevice))
            {
                AddPutEntry(bundle, _provenanceDevice);
            }

            AddProvenanceEntries(bundle);
        }

        return bundle;
    }

    /// <summary>
    /// Builds and returns two FHIR Bundles: a data bundle containing only the data resources and
    /// delete entries, and a provenance bundle containing the Provenance resource(s) and optionally
    /// the Device resource (if configured via <see cref="WithProvenance(Device, ResourceReference)"/>).
    /// The provenance bundle always uses <see cref="Bundle.BundleType.Transaction"/>.
    /// </summary>
    /// <returns>The data and provenance bundles.</returns>
    /// <exception cref="InvalidOperationException">
    /// Provenance has not been enabled via <see cref="WithProvenance(ResourceReference, ResourceReference)"/>
    /// or <see cref="WithProvenance(Device, ResourceReference)"/>, or
    /// <see cref="FailOnDuplicateEntries"/> is enabled and duplicate resource ids were found.
    /// </exception>
    public DataAndProvenanceBundles BuildWithSeparateProvenance()
    {
        if (!IsProvenanceEnabled)
        {
            throw new InvalidOperationException(
                "Provenance must be enabled via WithProvenance() before calling BuildWithSeparateProvenance()"
            );
        }

        var dataBundle = CreateBundle(_bundleType, _bundleId);
        AddDataEntries(dataBundle);

        // Use a SHA-256 hash of "provenance-" + the data bundle id for the provenance bundle id.
        // This keeps the id unique while ensuring it stays within FHIR id length/character
        // constraints.
        // We could also just use the data bundle id, but if both bundles are written to the same
        // topic in Kafka and that id is used as the key, it would cause issues on compaction.
        var provenanceBundle = CreateBundle(
            Bundle.BundleType.Transaction,
            _bundleId is null ? null : Sha256Hex($"provenance-{_bundleId}")
        );

        if (_provenanceDevice is not null)
        {
            AddPutEntry(provenanceBundle, _provenanceDevice);
        }

        AddProvenanceEntries(provenanceBundle);

        return new DataAndProvenanceBundles(dataBundle, provenanceBundle);
    }

    private static Bundle CreateBundle(Bundle.BundleType type, string? id) =>
        new() { Type = type, Id = id };

    /// <summary>
    /// Builds the <c>fullUrl</c> for a bundle entry from the resource's relative reference (e.g.
    /// <c>Patient/123</c>).
    ///
    /// <para>If a base URL has been configured via <see cref="WithFullUrlBase"/>, the fullUrl is
    /// <c>&lt;baseUrl&gt;/&lt;reference&gt;</c> - an absolute, hierarchical URL, as required by the
    /// FHIR spec. Other resources in the bundle can still reference this entry using the plain
    /// relative reference, since Bundle reference resolution rules match a relative reference
    /// against the tail of such a fullUrl.</para>
    ///
    /// <para>Otherwise, the relative reference itself is used, unchanged.</para>
    /// </summary>
    private string BuildFullUrl(string reference) =>
        _fullUrlBase is null ? reference : $"{_fullUrlBase}/{reference}";

    private void AddDataEntries(Bundle bundle)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);

        foreach (var resource in _resources)
        {
            var url = ReferenceUtils.CreateReferenceTo(resource).Reference!;

            if (_failOnDuplicateEntries && !seen.Add(url))
            {
                throw new InvalidOperationException($"Duplicate resource added: {url}");
            }

            AddPutEntry(bundle, resource, url);
        }

        foreach (var toDeleteReference in _resourcesToDelete)
        {
            bundle.Entry.Add(
                new Bundle.EntryComponent
                {
                    Request = new Bundle.RequestComponent
                    {
                        Method = Bundle.HTTPVerb.DELETE,
                        Url = toDeleteReference.Reference,
                    },
                }
            );
        }
    }

    private void AddProvenanceEntries(Bundle bundle)
    {
        if (_resources.Count > 0)
        {
            AddPutEntry(bundle, BuildCreateProvenance());
        }

        if (_resourcesToDelete.Count > 0)
        {
            AddPutEntry(bundle, BuildDeleteProvenance());
        }
    }

    private void AddPutEntry(Bundle bundle, Resource resource) =>
        AddPutEntry(bundle, resource, ReferenceUtils.CreateReferenceTo(resource).Reference!);

    private void AddPutEntry(Bundle bundle, Resource resource, string url) =>
        bundle.Entry.Add(
            new Bundle.EntryComponent
            {
                FullUrl = BuildFullUrl(url),
                Resource = resource,
                Request = new Bundle.RequestComponent { Method = Bundle.HTTPVerb.PUT, Url = url },
            }
        );

    private Provenance BuildCreateProvenance()
    {
        var (who, what) = RequireProvenance();
        var now = DateTimeOffset.UtcNow;

        var targets = _resources.Select(ReferenceUtils.CreateReferenceTo).ToList();

        if (_provenanceDevice is not null)
        {
            // If a device is configured as the provenance agent, we also want to include a
            // reference to it in the targets list.
            var deviceReference = ReferenceUtils.CreateReferenceTo(_provenanceDevice);
            if (!targets.Exists(target => target.Reference == deviceReference.Reference))
            {
                targets.Add(deviceReference);
            }
        }

        return new Provenance
        {
            Id = Sha256Hex($"create-{BuildProvenanceIdString(who, what)}"),
            Occurred = new FhirDateTime(now),
            Recorded = now,
            Target = targets,
            Activity = Concept(DataOperationSystem, "CREATE", "create"),
            Agent =
            {
                new Provenance.AgentComponent
                {
                    Type = Concept(ParticipantTypeSystem, "assembler", "Assembler"),
                    Role = { Concept(ParticipationTypeSystem, "AUT", "author") },
                    Who = who,
                },
            },
            Entity =
            {
                new Provenance.EntityComponent
                {
                    Role = Provenance.ProvenanceEntityRole.Source,
                    What = what,
                },
            },
        };
    }

    private Provenance BuildDeleteProvenance()
    {
        var (who, what) = RequireProvenance();
        var now = DateTimeOffset.UtcNow;

        var provenance = new Provenance
        {
            Id = Sha256Hex($"delete-{BuildProvenanceIdString(who, what)}"),
            Occurred = new FhirDateTime(now),
            Recorded = now,
            Target = [.. _resourcesToDelete],
            Activity = Concept(DataOperationSystem, "DELETE", "delete"),
            Agent =
            {
                new Provenance.AgentComponent
                {
                    Type = Concept(ParticipantTypeSystem, "performer", "Performer"),
                    Who = who,
                },
            },
        };

        foreach (var toDelete in _resourcesToDelete)
        {
            provenance.Entity.Add(
                new Provenance.EntityComponent
                {
                    Role = Provenance.ProvenanceEntityRole.Removal,
                    What = toDelete,
                }
            );
        }

        return provenance;
    }

    private (ResourceReference Who, ResourceReference What) RequireProvenance() =>
        _provenanceWho is { } who && _provenanceWhat is { } what
            ? (who, what)
            : throw new InvalidOperationException(
                "Provenance must be enabled via WithProvenance() before building a Provenance resource"
            );

    private static string BuildProvenanceIdString(ResourceReference who, ResourceReference what) =>
        $"{Describe(who, "who")}-{Describe(what, "what")}";

    /// <summary>
    /// Reduces a reference to the most specific stable string available - its literal reference,
    /// else its identifier value, else its display - for use in a deterministic Provenance id.
    /// </summary>
    private static string Describe(ResourceReference reference, string name)
    {
        if (!string.IsNullOrWhiteSpace(reference.Reference))
        {
            return reference.Reference;
        }

        if (reference.Identifier?.Value is { } identifierValue)
        {
            return identifierValue;
        }

        if (reference.Display is { } display)
        {
            return display;
        }

        throw new InvalidOperationException(
            $"Invalid provenance {name} reference. Either reference, identifier or display must be provided."
        );
    }

    private static CodeableConcept Concept(string system, string code, string display) =>
        new()
        {
            Coding =
            {
                new Coding
                {
                    System = system,
                    Code = code,
                    Display = display,
                },
            },
        };

    private static string Sha256Hex(string value) =>
        Convert.ToHexStringLower(SHA256.HashData(Encoding.UTF8.GetBytes(value)));
}

File: to-fhir-py/src/to_fhir/transaction_builder.py
"""Builder for creating FHIR transaction bundles."""

from __future__ import annotations

import hashlib
from datetime import UTC, datetime
from typing import NamedTuple, Self, cast

from fhir.resources.R4B.bundle import Bundle, BundleEntry, BundleEntryRequest
from fhir.resources.R4B.codeableconcept import CodeableConcept
from fhir.resources.R4B.coding import Coding
from fhir.resources.R4B.device import Device
from fhir.resources.R4B.provenance import Provenance, ProvenanceAgent, ProvenanceEntity
from fhir.resources.R4B.reference import Reference
from fhir.resources.R4B.resource import Resource

from to_fhir.reference_utils import create_reference_to

_DATA_OPERATION_SYSTEM = "http://terminology.hl7.org/CodeSystem/v3-DataOperation"
_PARTICIPANT_TYPE_SYSTEM = "http://terminology.hl7.org/CodeSystem/provenance-participant-type"
_PARTICIPATION_TYPE_SYSTEM = "http://terminology.hl7.org/CodeSystem/v3-ParticipationType"


class DataAndProvenanceBundles(NamedTuple):
    """A pair of FHIR bundles: one holding data resources, one holding provenance resources."""

    #: The bundle containing data resources (e.g. Patient, Observation).
    data_bundle: Bundle

    #: The bundle containing Provenance and optionally Device resources.
    provenance_bundle: Bundle


class TransactionBuilder:
    """Builder for FHIR transaction bundles, using the update-as-create approach by default."""

    def __init__(self) -> None:
        self._resources: list[Resource] = []
        self._resources_to_delete: list[Reference] = []
        self._bundle_type = "transaction"
        self._bundle_id: str | None = None
        self._full_url_base: str | None = None
        self._fail_on_duplicate_entries = False
        self._provenance_who: Reference | None = None
        self._provenance_what: Reference | None = None
        self._provenance_device: Device | None = None

    def with_type(self, bundle_type: str) -> Self:
        """Set the bundle type, e.g. ``"transaction"`` or ``"batch"``."""
        self._bundle_type = bundle_type
        return self

    def fail_on_duplicate_entries(self) -> Self:
        """Raise when multiple resources with the same id are added, once the bundle is built."""
        self._fail_on_duplicate_entries = True
        return self

    def add_entry(self, resource: Resource) -> Self:
        """Add a FHIR resource to the transaction bundle."""
        self._resources.append(resource)
        return self

    def add_entries(self, *resources: Resource) -> Self:
        """Add FHIR resources to the transaction bundle."""
        self._resources.extend(resources)
        return self

    def add_delete_entry(self, resource: Reference) -> Self:
        """Add a reference to a resource that should be deleted as part of the transaction."""
        self._resources_to_delete.append(resource)
        return self

    def add_delete_entries(self, *resources: Reference) -> Self:
        """Add references to resources that should be deleted as part of the transaction."""
        self._resources_to_delete.extend(resources)
        return self

    def with_id(self, bundle_id: str) -> Self:
        """Set the id of the bundle."""
        self._bundle_id = bundle_id
        return self

    def with_full_url_base(self, base_url: str) -> Self:
        """Set an absolute base URL used to build each entry's ``fullUrl``.

        Each entry's ``fullUrl`` becomes ``<base_url>/<ResourceType>/<id>``, making it absolute as
        FHIR requires. Plain ``ResourceType/id`` references used elsewhere in the bundle (e.g. in a
        ``Provenance.target``) still resolve correctly against it, since those match by the tail of
        a hierarchical ``fullUrl``. The base URL does not need to be a real, dereferenceable server
        endpoint.

        Args:
            base_url: An absolute base URL, e.g. ``https://example.org/fhir``. A trailing slash is
                optional.
        """
        self._full_url_base = base_url.rstrip("/")
        return self

    def with_provenance(self, who: Reference | Device, what: Reference) -> Self:
        """Include a Provenance resource in the bundle.

        If the bundle contains both delete and update/create entries, two Provenance resources are
        included. The ``Provenance.id`` is derived from the hash of ``who`` and ``what``.

        Args:
            who: The agent responsible for the transformation or deletion, as ``Provenance
                .agent.who``. This is typically the transformation service itself. Passing a
                ``Device`` adds that resource to the bundle and references it instead.
            what: A reference to the resource that is the source of the transformation, as
                ``Provenance.entity.what``.
        """
        if isinstance(who, Device):
            self._provenance_device = who
            self._provenance_who = create_reference_to(who)
        else:
            self._provenance_who = who

        self._provenance_what = what
        return self

    def build(self) -> Bundle:
        """Build a FHIR Bundle with the configured type.

        Raises:
            ValueError: :meth:`fail_on_duplicate_entries` is enabled and duplicate resource ids
                were found.
        """
        entries = self._data_entries()

        if self._provenance_enabled:
            device = self._provenance_device
            if device is not None and not any(resource is device for resource in self._resources):
                entries.append(self._put_entry(device))

            entries.extend(self._provenance_entries())

        return _bundle(self._bundle_type, self._bundle_id, entries)

    def build_with_separate_provenance(self) -> DataAndProvenanceBundles:
        """Build the data resources and the provenance resources as two separate bundles.

        The data bundle holds only the data resources and delete entries; the provenance bundle
        holds the Provenance resource(s) and, if configured via :meth:`with_provenance`, the
        Device. The provenance bundle is always of type ``transaction``.

        Raises:
            RuntimeError: Provenance has not been enabled via :meth:`with_provenance`.
            ValueError: :meth:`fail_on_duplicate_entries` is enabled and duplicate resource ids
                were found.
        """
        if not self._provenance_enabled:
            raise RuntimeError(
                "Provenance must be enabled via with_provenance() before calling "
                "build_with_separate_provenance()"
            )

        data_bundle = _bundle(self._bundle_type, self._bundle_id, self._data_entries())

        # Use a SHA-256 hash of "provenance-" + the data bundle id for the provenance bundle id.
        # This keeps the id unique while ensuring it stays within FHIR id length/character
        # constraints.
        # We could also just use the data bundle id, but if both bundles are written to the same
        # topic in Kafka and that id is used as the key, it would cause issues on compaction.
        provenance_bundle_id = (
            None if self._bundle_id is None else _sha256_hex(f"provenance-{self._bundle_id}")
        )

        entries: list[BundleEntry] = []
        if self._provenance_device is not None:
            entries.append(self._put_entry(self._provenance_device))

        entries.extend(self._provenance_entries())

        return DataAndProvenanceBundles(
            data_bundle, _bundle("transaction", provenance_bundle_id, entries)
        )

    @property
    def _provenance_enabled(self) -> bool:
        return self._provenance_who is not None and self._provenance_what is not None

    def _full_url(self, reference: str) -> str:
        """Build an entry's ``fullUrl`` from the resource's relative reference."""
        if self._full_url_base is None:
            return reference
        return f"{self._full_url_base}/{reference}"

    def _put_entry(self, resource: Resource, url: str | None = None) -> BundleEntry:
        url = url if url is not None else cast("str", create_reference_to(resource).reference)
        return BundleEntry(
            fullUrl=self._full_url(url),
            resource=resource,
            request=BundleEntryRequest(method="PUT", url=url),
        )

    def _data_entries(self) -> list[BundleEntry]:
        entries: list[BundleEntry] = []
        seen: set[str] = set()

        for resource in self._resources:
            url = cast("str", create_reference_to(resource).reference)

            if self._fail_on_duplicate_entries and url in seen:
                raise ValueError(f"Duplicate resource added: {url}")

            seen.add(url)
            entries.append(self._put_entry(resource, url))

        entries.extend(
            BundleEntry(request=BundleEntryRequest(method="DELETE", url=reference.reference))
            for reference in self._resources_to_delete
        )

        return entries

    def _provenance_entries(self) -> list[BundleEntry]:
        entries: list[BundleEntry] = []

        if self._resources:
            entries.append(self._put_entry(self._build_create_provenance()))

        if self._resources_to_delete:
            entries.append(self._put_entry(self._build_delete_provenance()))

        return entries

    def _require_provenance(self) -> tuple[Reference, Reference]:
        if self._provenance_who is None or self._provenance_what is None:
            raise RuntimeError(
                "Provenance must be enabled via with_provenance() before building a "
                "Provenance resource"
            )
        return self._provenance_who, self._provenance_what

    def _build_create_provenance(self) -> Provenance:
        who, what = self._require_provenance()
        now = datetime.now(UTC)

        targets = [create_reference_to(resource) for resource in self._resources]

        if self._provenance_device is not None:
            # If a device is configured as the provenance agent, we also want to include a
            # reference to it in the targets list.
            device_reference = create_reference_to(self._provenance_device)
            if all(target.reference != device_reference.reference for target in targets):
                targets.append(device_reference)

        return Provenance(
            id=_sha256_hex(f"create-{_provenance_id_string(who, what)}"),
            occurredDateTime=now,
            recorded=now,
            target=targets,
            activity=_concept(_DATA_OPERATION_SYSTEM, "CREATE", "create"),
            agent=[
                ProvenanceAgent(
                    type=_concept(_PARTICIPANT_TYPE_SYSTEM, "assembler", "Assembler"),
                    role=[_concept(_PARTICIPATION_TYPE_SYSTEM, "AUT", "author")],
                    who=who,
                )
            ],
            entity=[ProvenanceEntity(role="source", what=what)],
        )

    def _build_delete_provenance(self) -> Provenance:
        who, what = self._require_provenance()
        now = datetime.now(UTC)

        return Provenance(
            id=_sha256_hex(f"delete-{_provenance_id_string(who, what)}"),
            occurredDateTime=now,
            recorded=now,
            target=list(self._resources_to_delete),
            activity=_concept(_DATA_OPERATION_SYSTEM, "DELETE", "delete"),
            agent=[
                ProvenanceAgent(
                    type=_concept(_PARTICIPANT_TYPE_SYSTEM, "performer", "Performer"),
                    who=who,
                )
            ],
            entity=[
                ProvenanceEntity(role="removal", what=to_delete)
                for to_delete in self._resources_to_delete
            ],
        )


def _bundle(bundle_type: str, bundle_id: str | None, entries: list[BundleEntry]) -> Bundle:
    return Bundle(type=bundle_type, id=bundle_id, entry=entries or None)


def _concept(system: str, code: str, display: str) -> CodeableConcept:
    return CodeableConcept(coding=[Coding(system=system, code=code, display=display)])


def _sha256_hex(value: str) -> str:
    return hashlib.sha256(value.encode()).hexdigest()


def _provenance_id_string(who: Reference, what: Reference) -> str:
    return f"{_describe(who, 'who')}-{_describe(what, 'what')}"


def _describe(reference: Reference, name: str) -> str:
    """Reduce a reference to the most specific stable string available.

    That is its literal reference, else its identifier value, else its display - for use in a
    deterministic Provenance id.
    """
    literal = reference.reference
    if literal and literal.strip():
        return literal

    if reference.identifier is not None and reference.identifier.value is not None:
        return reference.identifier.value

    if reference.display is not None:
        return reference.display

    raise ValueError(
        f"Invalid provenance {name} reference. "
        f"Either reference, identifier or display must be provided."
    )
